{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore sample-level QC results\n",
    "\n",
    "In this notebook, we explore sample-level QC results generated by [Sample-Level-QC.Rmd](https://github.com/verilylifesciences/variant-qc/blob/master/R/Sample-Level-QC.Rmd) and examine the outliers.\n",
    "\n",
    "<div class='alert alert-block alert-info'>\n",
    "<b>How to use this notebook:</b><p>\n",
    "    \n",
    "<a href=\"https://github.com/verilylifesciences/variant-qc/blob/master/R/Sample-Level-QC.Rmd\">RMarkdown</a> was used to run a set of analyses and generate a canned report. If you have questions about the details of the analyses, see the canned reports for details. This notebook reads in those previously computed results, and re-renders 'small multiples' of the results.<p>\n",
    "    \n",
    "In this notebook you can:<ul>    \n",
    "        <li>Use <a href=\"https://papermill.readthedocs.io/en/latest/usage-parameterize.html\">Papermill parameters</a> to select which dataset you'd like to explore. Options include:<ul>\n",
    "            <li><a href=\"https://www.simonsfoundation.org/simons-genome-diversity-project/\">The Simons Genome Diversity Project</a>\n",
    "            <li><a href=\"https://www.internationalgenome.org/category/phase-3/\">1000 Geomes Phase 3 Variants</a>\n",
    "            <li>Deep Variant Platinum Genomes\n",
    "            </ul>\n",
    "        <li>Modify the plots\n",
    "        <li>Add new plots\n",
    "        <li>Change the QC cutoff thresholds when examining outliers\n",
    "        <li>Use updated sample information\n",
    "        <li>Use more features from the sample information\n",
    "        <li>Choose a custom output filename for the QC problems summary\n",
    "    </ul><p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lapply(c('devtools', 'tidyverse', 'bigrquery', 'scales', 'skimr'),\n",
    "       function(pkg_name) { if(! pkg_name %in% installed.packages()) { install.packages(pkg_name)} } )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(bigrquery)\n",
    "library(scales)\n",
    "library(skimr)\n",
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AUTOSOMES <- c(paste('chr', c(as.character(seq(1, 22))), sep = ''),\n",
    "               as.character(seq(1, 22)))\n",
    "ALLOSOMES <- c('chrX', 'X', 'chrY', 'Y')\n",
    "MITOCHONDRIA <- c('chrM', 'M', 'MT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEFAULT_PLOT_HEIGHT <- 7.5\n",
    "options(repr.plot.width = 16, repr.plot.height = DEFAULT_PLOT_HEIGHT)\n",
    "theme_set(theme_bw(base_size = 16))\n",
    "update_geom_defaults('point', list(size = 2))\n",
    "\n",
    "#' Returns a data frame with a y position and a label, for use annotating ggplot boxplots.\n",
    "#'\n",
    "#' @param d A data frame.\n",
    "#' @return A data frame with column y as max and column label as length.\n",
    "#'\n",
    "get_boxplot_fun_data <- function(d) {\n",
    "  return(data.frame(y = max(d), label = stringr::str_c('N = ', length(d))))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCS paths and BigQuery queries for datasets\n",
    "DATASETS <- c('simons', 'thousand_genomes', 'platinum_genomes')\n",
    "\n",
    "CHROM_RESULTS_PATHS <- list(simons = 'gs://genomics-public-data/simons-genome-diversity-project/reports/Simons_Genome_Diversity_Project_sample_reference_results.csv',\n",
    "                            platinum_genomes = 'gs://genomics-public-data/platinum-genomes/reports/DeepVariant_Platinum_Genomes_sample_reference_results.csv',\n",
    "                            thousand_genomes = 'gs://genomics-public-data/1000-genomes-phase-3/reports/1000_Genomes_phase_3_sample_reference_results.csv')\n",
    "\n",
    "GENOME_RESULTS_PATHS <- list(simons = 'gs://genomics-public-data/simons-genome-diversity-project/reports/Simons_Genome_Diversity_Project_sample_results.csv',\n",
    "                             platinum_genomes = 'gs://genomics-public-data/platinum-genomes/reports/DeepVariant_Platinum_Genomes_sample_results.csv',\n",
    "                             thousand_genomes = 'gs://genomics-public-data/1000-genomes-phase-3/reports/1000_Genomes_phase_3_sample_results.csv')\n",
    "\n",
    "simons_sample_info_query <- '\n",
    "    SELECT\n",
    "      id_from_vcf AS name,\n",
    "      \"Illumina\" AS sequencing_platform,\n",
    "      sex,\n",
    "      region AS ancestry\n",
    "    FROM\n",
    "      `bigquery-public-data.human_genome_variants.simons_genome_diversity_project_sample_attributes`\n",
    "    '\n",
    "\n",
    "thousand_genomes_sample_info_query <- '\n",
    "    SELECT\n",
    "      Sample AS name,\n",
    "      \"Illumina\" AS sequencing_platform,\n",
    "      Gender AS sex,\n",
    "      Super_Population AS ancestry\n",
    "    FROM\n",
    "      `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`\n",
    "    '\n",
    "\n",
    "SAMPLE_INFO_QUERIES <- list(simons = simons_sample_info_query,\n",
    "                        thousand_genomes = thousand_genomes_sample_info_query,\n",
    "                        platinum_genomes = thousand_genomes_sample_info_query) # Use 1000 genomes for DV platinum"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Papermill Parameters\n",
    "\n",
    "This next cell contains [Papermill parameters](https://papermill.readthedocs.io/en/latest/usage-parameterize.html). You'll notice it has a `parameters` cell tag. You can set cell tags by using the `View` > `Cell Toolbar` > `Tags` menu option.\n",
    "\n",
    "Use the `DATASET` parameter to select which dataset you'd like to examine:\n",
    "\n",
    "* `simons` for the Simons Genome Diversity Project\n",
    "* `thousand_genomes` for the 1000 Genomes Phase 3 Release\n",
    "* `platinum_genomes` for the Deep Variant platinum genomes\n",
    "\n",
    "Use the `CSV_OUTPUT_FILE_NAME` parameter to name your CSV output file. If you leave it unset, it will default to `<DATASET>_problem_summary.csv`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "BILLING_PROJECT_ID <- Sys.getenv('GOOGLE_PROJECT')\n",
    "DATASET <- 'simons'\n",
    "CSV_OUTPUT_FILE_NAME <- NULL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Validate Papermill parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (!(DATASET %in% DATASETS)) {\n",
    "    stop(str_glue(DATASET, ' is not a valid dataset name. Valid names are: [', \n",
    "                  paste(DATASETS, collapse = ', '), ']'))\n",
    "}\n",
    "\n",
    "if (is.null(CSV_OUTPUT_FILE_NAME)) {\n",
    "    CSV_OUTPUT_FILE_NAME <- str_glue(DATASET, '_problem_summary.csv')\n",
    "}\n",
    "\n",
    "print(DATASET)\n",
    "print(CSV_OUTPUT_FILE_NAME)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieve sample information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query sample info for `DATASET` from BigQuery\n",
    "sample_info <- bq_table_download(bq_project_query(BILLING_PROJECT_ID, SAMPLE_INFO_QUERIES[[DATASET]]))\n",
    "\n",
    "print(skim(sample_info))\n",
    "dim(sample_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieve previously computed results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_results <- read_csv(pipe(str_glue('gsutil cat ', GENOME_RESULTS_PATHS[[DATASET]])))\n",
    "genome_results <- genome_results[,!grepl('X1', colnames(genome_results))]  # Drop row name column, if present.\n",
    "\n",
    "print(skim(genome_results))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_results <- read_csv(pipe(str_glue('gsutil cat ', CHROM_RESULTS_PATHS[[DATASET]])))\n",
    "chrom_results <- chrom_results[,!grepl('X1', colnames(chrom_results))]  # Drop row name column, if present.\n",
    "\n",
    "print(skim(chrom_results))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined_genome_results <- inner_join(genome_results, sample_info)\n",
    "\n",
    "dim(joined_genome_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined_chrom_results <- inner_join(chrom_results, sample_info)\n",
    "\n",
    "dim(joined_chrom_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample information "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info %>%\n",
    "  group_by(ancestry, sex) %>%\n",
    "  summarize(\n",
    "    count = n(),\n",
    "    proportion = n() / nrow(.)\n",
    "  ) %>%\n",
    "ggplot(aes(x = ancestry, y = count, fill = ancestry)) +\n",
    "  geom_bar(stat = 'identity', position = 'dodge') +\n",
    "  geom_text(aes(label = str_glue('N = {count} ({percent(proportion)})'))) +\n",
    "  facet_wrap( ~ sex, ncol = 1) +\n",
    "  theme(axis.text.x = element_text(angle = 50, hjust = 1)) +\n",
    "  ggtitle('Sample count and proportion by ancestry and sex')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info %>%\n",
    "  group_by(sequencing_platform, sex) %>%\n",
    "  summarize(\n",
    "    count = n(),\n",
    "    proportion = n() / nrow(.)\n",
    "  ) %>%\n",
    "ggplot(aes(x = sequencing_platform, y = count, fill = sequencing_platform)) +\n",
    "  geom_bar(stat = 'identity', position = 'dodge') +\n",
    "  geom_text(aes(label = str_glue('N = {count} ({percent(proportion)})'))) +\n",
    "  coord_cartesian(clip = 'off') +\n",
    "  facet_wrap(. ~ sex, ncol = 1) +\n",
    "  theme(axis.text.x = element_text(angle = 50, hjust = 1)) +\n",
    "  ggtitle('Sample count and proportion by sequencing platform and sex')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variant call rate\n",
    "\n",
    "## Per-chromosome variant call rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined_chrom_results %>%\n",
    "  filter(reference_name %in% c(AUTOSOMES, ALLOSOMES, MITOCHONDRIA)) %>%\n",
    "  mutate(reference_name = parse_factor(reference_name, levels = c(AUTOSOMES, ALLOSOMES, MITOCHONDRIA))) %>%\n",
    "ggplot(aes(y = number_of_calls, x = reference_name)) +\n",
    "  geom_boxplot() +\n",
    "  scale_y_continuous(labels = comma) +\n",
    "  facet_wrap(. ~ sequencing_platform, ncol = 1, scales = 'free_y') +\n",
    "  ylab('Number of Variant Calls') +\n",
    "  xlab('Chromosome') +\n",
    "  ggtitle('Box plot: Count of variant calls per genome by chromosome') +\n",
    "  theme(axis.text.x = element_text(angle = 50, hjust = 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display per-chromosome outliers "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compute per chromosome mean and sd call count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_calls_summary <- joined_chrom_results %>%\n",
    "    group_by(reference_name, sequencing_platform) %>%\n",
    "    summarize(\n",
    "        number_of_calls_mean = mean(number_of_calls),\n",
    "        number_of_calls_sd = sd(number_of_calls),\n",
    "    )\n",
    "\n",
    "head(chrom_calls_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "low_chrom_calls_sd_multiplier <- 3\n",
    "\n",
    "(chrom_calls_outliers <- joined_chrom_results %>%\n",
    "    inner_join(chrom_calls_summary) %>%\n",
    "    filter(number_of_calls < number_of_calls_mean - (low_chrom_calls_sd_multiplier * number_of_calls_sd)) %>%\n",
    "    arrange(reference_name, number_of_calls))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "problems <- chrom_calls_outliers %>%\n",
    "                 group_by(name) %>%\n",
    "                 summarize(\n",
    "                     fail = str_glue('per chrom variant count is at least {low_chrom_calls_sd_multiplier}',\n",
    "                                     ' standard deviations below the platform mean for ',\n",
    "                                     str_c(reference_name, collapse = ',')))\n",
    "\n",
    "dim(problems)\n",
    "head(problems)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Per-genome variant call rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_call_rate_summary <- joined_chrom_results %>%\n",
    "  group_by(name, sex, ancestry) %>%\n",
    "  summarize(n = sum(number_of_calls))\n",
    "\n",
    "p <- genome_call_rate_summary %>%\n",
    "  ggplot() +\n",
    "  geom_point(aes(x = name, y = n, color = ancestry, shape = sex)) +\n",
    "  scale_x_discrete(expand = c(0.05, 1)) +\n",
    "  scale_y_continuous(labels = comma) +\n",
    "  xlab('Sample') +\n",
    "  ylab('Number of Calls') +\n",
    "  ggtitle('Scatter Plot: Count of Calls Per Genome') +\n",
    "  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),\n",
    "            panel.grid.major.x = element_blank())\n",
    "if (nrow(genome_call_rate_summary) <= 20) {\n",
    "  p + theme(axis.text.x = element_text(angle = 50, hjust = 1))\n",
    "} else {\n",
    "  p + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),\n",
    "            panel.grid.major.x = element_blank())\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_call_rate_summary <- joined_chrom_results %>%\n",
    "  group_by(name, sex, sequencing_platform) %>%\n",
    "  summarize(n = sum(number_of_calls))\n",
    "\n",
    "p <- genome_call_rate_summary %>%\n",
    "  ggplot() +\n",
    "  geom_point(aes(x = name, y = n, color = sequencing_platform, shape = sex)) +\n",
    "  scale_x_discrete(expand = c(0.05, 1)) +\n",
    "  scale_y_continuous(labels = comma) +\n",
    "  xlab('Sample') +\n",
    "  ylab('Number of Calls') +\n",
    "  ggtitle('Scatter Plot: Count of Calls Per Genome') +\n",
    "  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),\n",
    "            panel.grid.major.x = element_blank())\n",
    "if (nrow(genome_call_rate_summary) <= 20) {\n",
    "  p + theme(axis.text.x = element_text(angle = 50, hjust = 1))\n",
    "} else {\n",
    "  p + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),\n",
    "            panel.grid.major.x = element_blank())\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display per-genome outliers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_calls_summary <- genome_call_rate_summary %>%\n",
    "    group_by(sequencing_platform) %>%\n",
    "    summarize(\n",
    "        genome_calls_mean = mean(n),\n",
    "        genome_calls_sd = sd(n),\n",
    "    )\n",
    "\n",
    "head(genome_calls_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "low_genome_calls_sd_multiplier <- 3\n",
    "\n",
    "(genome_calls_outliers <- genome_call_rate_summary %>%\n",
    "    inner_join(genome_calls_summary) %>%\n",
    "    filter(n < genome_calls_mean - (low_genome_calls_sd_multiplier * genome_calls_sd)) %>%\n",
    "    arrange(n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "problems <- rbind(\n",
    "    tibble(name = unique(genome_calls_outliers$name),\n",
    "           fail = str_glue('per genome variant count is at least {low_genome_calls_sd_multiplier}',\n",
    "                           ' standard deviations below the platform mean')),\n",
    "    problems)\n",
    "\n",
    "dim(problems)\n",
    "head(problems)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Missingness rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined_genome_results %>%\n",
    "ggplot(aes(y = missingness_rate, x = ancestry, fill = ancestry)) +\n",
    "  geom_boxplot() +\n",
    "  stat_summary(fun.data = get_boxplot_fun_data, geom = 'text', size = 3,\n",
    "               position = position_dodge(width = 0.9), vjust = -0.5) +\n",
    "  scale_y_continuous(labels = percent_format()) +\n",
    "  facet_wrap(. ~ sequencing_platform, ncol = 1, scales = 'free_y') +\n",
    "  ylab('Missingness Rate') +\n",
    "  xlab('Ancestry') +\n",
    "  ggtitle('Genome-Specific Missingness') +\n",
    "  theme(axis.text.x = element_text(angle = 50, hjust = 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- joined_genome_results %>%\n",
    "  ggplot(aes(x = name, y = missingness_rate, color = sex)) +\n",
    "  geom_point() +\n",
    "  scale_x_discrete(expand = c(0.05, 1)) +\n",
    "  scale_y_continuous(limits = c(0, NA), labels = percent_format()) +\n",
    "  facet_wrap(. ~ sequencing_platform, ncol = 1, scales = 'free') +\n",
    "  xlab('Sample') +\n",
    "  ylab('Missingness Rate') +\n",
    "  ggtitle('Scatter Plot: Genome-Specific Missingness')\n",
    "if (nrow(joined_genome_results) <= 20) {\n",
    "  p + theme(axis.text.x = element_text(angle = 50, hjust = 1))\n",
    "} else {\n",
    "  p + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),\n",
    "            panel.grid.major.x = element_blank())\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display outliers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(missingness_summary <- joined_genome_results %>%\n",
    "    group_by(sequencing_platform) %>%\n",
    "    summarize(\n",
    "        missingness_rate_mean = mean(missingness_rate),\n",
    "        missingness_rate_sd = sd(missingness_rate),\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "low_missingness_sd_multiplier <- 3\n",
    "\n",
    "(low_missingness_outliers <- joined_genome_results %>%\n",
    "     inner_join(missingness_summary) %>%\n",
    "     filter(\n",
    "         missingness_rate < missingness_rate_mean - (low_missingness_sd_multiplier * missingness_rate_sd)\n",
    "     ) %>%\n",
    "     arrange(missingness_rate))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_missingness_sd_multiplier <- 3\n",
    "\n",
    "(high_missingness_outliers <- joined_genome_results %>%\n",
    "     inner_join(missingness_summary) %>%\n",
    "     filter(missingness_rate > missingness_rate_mean + (high_missingness_sd_multiplier * missingness_rate_sd)) %>%\n",
    "     arrange(desc(missingness_rate)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "problems <- rbind(\n",
    "    tibble(name = unique(high_missingness_outliers$name),\n",
    "           fail = str_glue('per genome missingness rate is at least {high_missingness_sd_multiplier}',\n",
    "                           ' standard deviations above the platform mean')),\n",
    "    tibble(name = unique(low_missingness_outliers$name),\n",
    "           fail = str_glue('per genome missingness rate is at least {low_missingness_sd_multiplier}',\n",
    "                           ' standard deviations below the platform mean')),\n",
    "    problems)\n",
    "\n",
    "dim(problems)\n",
    "head(problems)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Singleton rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined_genome_results %>%\n",
    "  ggplot(aes(y = private_variant_count, x = ancestry, fill = ancestry)) +\n",
    "  geom_boxplot() +\n",
    "  stat_summary(fun.data = get_boxplot_fun_data, geom = 'text',\n",
    "               position = position_dodge(width = 0.9), vjust = -0.5) +\n",
    "  facet_wrap(. ~ sequencing_platform, ncol = 1, scales = 'free_y') +\n",
    "  scale_y_continuous(labels = comma, expand = c(0.3, 0)) +\n",
    "  ylab('Number of Singletons') +\n",
    "  xlab('Ancestry') +\n",
    "  ggtitle('Box plot: Count of singletons per genome by ancestry') +\n",
    "  theme(axis.text.x = element_text(angle = 50, hjust = 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- joined_genome_results %>%\n",
    "  ggplot(aes(x = name, y = private_variant_count, color = sex)) +\n",
    "  geom_point() +\n",
    "  scale_x_discrete(expand = c(0.05, 1)) +\n",
    "  scale_y_log10(labels = comma) +\n",
    "  facet_wrap(. ~ sequencing_platform, ncol = 1, scales = 'free') +\n",
    "  xlab('Sample') +\n",
    "  ylab('Number of Singletons (log scale)') +\n",
    "  ggtitle('Scatter Plot: Count of Singletons Per Genome')\n",
    "if (nrow(joined_genome_results) <= 20) {\n",
    "  p + theme(axis.text.x = element_text(angle = 50, hjust = 1))\n",
    "} else {\n",
    "  p + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),\n",
    "            panel.grid.major.x = element_blank())\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display outliers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(singleton_summary <- joined_genome_results %>%\n",
    "     group_by(sequencing_platform, ancestry) %>%\n",
    "     summarize(\n",
    "         singleton_count_mean = mean(private_variant_count),\n",
    "         singleton_count_sd = sd(private_variant_count),\n",
    "     ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_singleton_sd_multiplier <- 3\n",
    "\n",
    "(high_singleton_outliers <- joined_genome_results %>%\n",
    "     inner_join(singleton_summary) %>%\n",
    "     filter(private_variant_count > singleton_count_mean + (high_singleton_sd_multiplier * singleton_count_sd)) %>%\n",
    "     arrange(desc(private_variant_count)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "problems <- rbind(\n",
    "    tibble(name = high_singleton_outliers$name,\n",
    "           fail = str_glue('singleton count at least {high_singleton_sd_multiplier}',\n",
    "           ' standard deviations above the ancestry and platform mean')),\n",
    "    problems)\n",
    "\n",
    "dim(problems)\n",
    "head(problems)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Heterozygosity rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot(joined_genome_results,\n",
    "       aes(y = heterozygous_variant_count, x = ancestry, fill = ancestry)) +\n",
    "  geom_boxplot() +\n",
    "  stat_summary(fun.data = get_boxplot_fun_data, geom = 'text',\n",
    "               position = position_dodge(width = 0.9), vjust = -0.5) +\n",
    "  scale_y_continuous(labels = comma, expand = c(0.3, 0)) +\n",
    "  facet_wrap(. ~ sequencing_platform, ncol = 1, scales = 'free_y') +\n",
    "  ylab('Number of Heterozyous Variants') +\n",
    "  xlab('Ancestry') +\n",
    "  ggtitle('Box plot: Count of heterozygous variants per genome by ancestry') +\n",
    "  theme(axis.text.x = element_text(angle = 50, hjust = 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(joined_genome_results) +\n",
    "  geom_point(aes(x = name, y = heterozygous_variant_count, color = sex)) +\n",
    "  scale_x_discrete(expand = c(0.05, 1)) +\n",
    "  scale_y_continuous(labels = comma) +\n",
    "  facet_wrap(. ~ sequencing_platform, ncol = 1, scales = 'free') +\n",
    "  xlab('Sample') +\n",
    "  ylab('Number of Heterozygous Variants') +\n",
    "  ggtitle('Scatter Plot: Count of Heterozygous Variants Per Genome')\n",
    "if (nrow(joined_genome_results) <= 20) {\n",
    "  p + theme(axis.text.x = element_text(angle = 50, hjust = 1))\n",
    "} else {\n",
    "  p + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),\n",
    "            panel.grid.major.x = element_blank())\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display outliers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(heterozygous_variant_count_summary <- joined_genome_results %>%\n",
    "    group_by(sequencing_platform, ancestry) %>%\n",
    "    summarize(\n",
    "        heterozygous_variant_count_mean = mean(heterozygous_variant_count),\n",
    "        heterozygous_variant_count_sd = sd(heterozygous_variant_count),\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "low_het_count_sd_multiplier <- 3\n",
    "\n",
    "(low_heterozygous_variant_count_outliers <- joined_genome_results %>%\n",
    "     inner_join(heterozygous_variant_count_summary) %>%\n",
    "     filter(heterozygous_variant_count\n",
    "            < heterozygous_variant_count_mean - (low_het_count_sd_multiplier * heterozygous_variant_count_sd)\n",
    "     ) %>%\n",
    "     arrange(heterozygous_variant_count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_het_count_sd_multiplier <- 3\n",
    "\n",
    "(high_heterozygous_variant_count_outliers <- joined_genome_results %>%\n",
    "     inner_join(heterozygous_variant_count_summary) %>%\n",
    "     filter(heterozygous_variant_count\n",
    "            > heterozygous_variant_count_mean + (high_het_count_sd_multiplier * heterozygous_variant_count_sd)\n",
    "     ) %>%\n",
    "     arrange(desc(heterozygous_variant_count)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "problems <- rbind(\n",
    "    tibble(name = low_heterozygous_variant_count_outliers$name,\n",
    "           fail = str_glue('heterozygous variant count at least {low_het_count_sd_multiplier}',\n",
    "                           ' standard deviations below the ancestry and platform mean')),\n",
    "    tibble(name = high_heterozygous_variant_count_outliers$name,\n",
    "           fail = str_glue('heterozygous variant count at least {high_het_count_sd_multiplier}',\n",
    "                           ' standard deviations above the ancestry and platform mean')),\n",
    "    problems)\n",
    "\n",
    "dim(problems)\n",
    "head(problems)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homozygosity rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined_chrom_results %>%\n",
    "  filter(reference_name %in% c(AUTOSOMES, ALLOSOMES, MITOCHONDRIA)) %>%\n",
    "  mutate(reference_name = parse_factor(reference_name, levels = c(AUTOSOMES, ALLOSOMES, MITOCHONDRIA))) %>%\n",
    "  ggplot(aes(y = F, x = reference_name, color = sex)) +\n",
    "  geom_boxplot() +\n",
    "  facet_grid(sex ~ .) +\n",
    "  ylab('Fraction of Homozygous Variants') +\n",
    "  xlab('Reference Name') +\n",
    "  ggtitle('Fraction of Homozygous Variants Per Genome') +\n",
    "  theme(axis.text.x = element_text(angle = 50, hjust = 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "head(fraction_homozygous_summary <- joined_chrom_results %>%\n",
    "    group_by(reference_name) %>%\n",
    "    summarize(\n",
    "        F_mean = mean(F),\n",
    "        F_sd = sd(F),\n",
    "    ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display per-chromosome outliers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "low_fraction_homozygous_sd_multiplier <- 4\n",
    "\n",
    "(low_hom_outliers <- joined_chrom_results %>%\n",
    "    inner_join(fraction_homozygous_summary) %>%\n",
    "    filter(F < F_mean - (low_fraction_homozygous_sd_multiplier * F_sd)) %>%\n",
    "    arrange(sex, reference_name, F))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_fraction_homozygous_sd_multiplier <- 4\n",
    "\n",
    "(high_hom_outliers <- joined_chrom_results %>%\n",
    "    inner_join(fraction_homozygous_summary) %>%\n",
    "    filter(F > F_mean + (high_fraction_homozygous_sd_multiplier * F_sd)) %>%\n",
    "    arrange(sex, reference_name, desc(F)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "problems <- rbind(\n",
    "    low_hom_outliers %>%\n",
    "    group_by(name) %>%\n",
    "    summarize(fail = str_glue('per chrom homozygosity is at least {low_fraction_homozygous_sd_multiplier}',\n",
    "                               ' standard deviations below the mean ',\n",
    "                              str_c(reference_name, collapse = ','))),\n",
    "    high_hom_outliers %>%\n",
    "    group_by(name) %>%\n",
    "    summarize(fail = str_glue('per chrom homozygosity is at least {high_fraction_homozygous_sd_multiplier}',\n",
    "                               ' standard deviations above the mean ',\n",
    "                              str_c(reference_name, collapse = ','))),\n",
    "    problems)\n",
    "\n",
    "dim(problems)\n",
    "head(problems)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ti/Tv ratio per chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined_chrom_results %>%\n",
    "  filter(reference_name %in% c(AUTOSOMES)) %>%\n",
    "  mutate(reference_name = parse_factor(reference_name, levels = c(AUTOSOMES))) %>%\n",
    "  ggplot(aes(y = titv, x = reference_name, color = sex)) +\n",
    "  geom_boxplot() +\n",
    "  ylab('Ti/Tv ratio') +\n",
    "  xlab('Chromosome') +\n",
    "  ggtitle('Ti/Tv ratio per genome') +\n",
    "  theme(axis.text.x = element_text(angle = 50, hjust = 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display per-chromosome outliers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "titv_summary <- joined_chrom_results %>%\n",
    "    group_by(reference_name) %>%\n",
    "    summarize(\n",
    "        titv_mean = mean(titv),\n",
    "        titv_sd = sd(titv),\n",
    "    )\n",
    "\n",
    "head(titv_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "low_titv_sd_multiplier <- 4\n",
    "\n",
    "(low_titv_outliers <- joined_chrom_results %>%\n",
    "    inner_join(titv_summary) %>%\n",
    "    filter(titv < titv_mean - (low_titv_sd_multiplier * titv_sd)) %>%\n",
    "    arrange(reference_name, titv))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_titv_sd_multiplier <- 4\n",
    "\n",
    "(high_titv_outliers <- joined_chrom_results %>%\n",
    "    inner_join(titv_summary) %>%\n",
    "    filter(titv > titv_mean + (high_titv_sd_multiplier * titv_sd)) %>%\n",
    "    arrange(reference_name, desc(titv)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "problems <- rbind(\n",
    "    low_titv_outliers %>%\n",
    "    group_by(name) %>%\n",
    "    summarize(fail = str_glue('per chrom ti/tv is at least {low_titv_sd_multiplier} standard deviations below the mean for ',\n",
    "                              str_c(reference_name, collapse = ','))),\n",
    "    high_titv_outliers %>%\n",
    "    group_by(name) %>%\n",
    "    summarize(fail = str_glue('per chrom ti/tv is at least {high_titv_sd_multiplier} standard deviations above the mean for ',\n",
    "                              str_c(reference_name, collapse = ','))),\n",
    "    problems)\n",
    "\n",
    "dim(problems)\n",
    "head(problems)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sex inference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot(joined_genome_results, aes(x = sex, y = perct_het_alt_in_snvs, fill = sex)) +\n",
    "  geom_boxplot() +\n",
    "  stat_summary(fun.data = get_boxplot_fun_data, geom = 'text', position = position_dodge(width = 0.9), vjust = -0.5) +\n",
    "  scale_y_continuous(labels = percent_format()) +\n",
    "  xlab('Sex') +\n",
    "  ylab('Heterozygosity Rate ') +\n",
    "  ggtitle('Box Plot: Heterozygosity Rate on the X Chromosome')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(joined_genome_results) +\n",
    "  geom_point(aes(x = name, y = perct_het_alt_in_snvs, color = sex)) +\n",
    "  scale_x_discrete(expand = c(0.05, 1)) +\n",
    "  scale_y_continuous(labels = percent_format()) +\n",
    "  facet_wrap(. ~ sequencing_platform, ncol = 1, scales = 'free') +\n",
    "  xlab('Sample') +\n",
    "  ylab('Heterozygosity Rate ') +\n",
    "  ggtitle('Scatter Plot: Heterozygosity Rate on the X Chromosome')\n",
    "if (nrow(joined_genome_results) <= 20) {\n",
    "  p + theme(axis.text.x = element_text(angle = 50, hjust = 1))\n",
    "} else {\n",
    "  p + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),\n",
    "            panel.grid.major.x = element_blank())\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display outliers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(m_outliers <- joined_genome_results %>%\n",
    "    filter(!sex %in% c('F', 'FEMALE', 'female') & perct_het_alt_in_snvs > 0.4) %>%\n",
    "    arrange(desc(perct_het_alt_in_snvs))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(f_outliers <- joined_genome_results %>%\n",
    "    filter(! sex %in% c('M', 'MALE', 'male') & perct_het_alt_in_snvs < 0.4) %>%\n",
    "    arrange(perct_het_alt_in_snvs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "problems <- rbind(\n",
    "    tibble(name = m_outliers$name, fail = str_glue('sample info indicates {m_outliers$sex}, genome indicates female')),\n",
    "    tibble(name = f_outliers$name, fail = str_glue('sample info indicates {f_outliers$sex}, genome indicates male')),\n",
    "    problems)\n",
    "\n",
    "dim(problems)\n",
    "head(problems)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem summary\n",
    "\n",
    "Group all the 'problems' together by sample and emit a CSV report."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(problem_summary <- problems %>%\n",
    "    group_by(name) %>%\n",
    "    summarize(\n",
    "        issues = str_c(fail, collapse = ';\\n '),\n",
    "        issue_count = n()\n",
    "    ) %>%\n",
    "    arrange(desc(issue_count), name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "write_csv(problem_summary, file = CSV_OUTPUT_FILE_NAME)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Provenance "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "devtools::session_info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copyright 2019 Verily Life Sciences LLC. All rights reserved.\n",
    "\n",
    "Licensed under the Apache License, Version 2.0 (the 'License');\n",
    "you may not use this file except in compliance with the License.\n",
    "You may obtain a copy of the License at\n",
    "\n",
    "     http://www.apache.org/licenses/LICENSE-2.0\n",
    "\n",
    "Unless required by applicable law or agreed to in writing, software\n",
    "distributed under the License is distributed on an 'AS IS' BASIS,\n",
    "WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
    "See the License for the specific language governing permissions and\n",
    "limitations under the License."
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "environment": {
   "name": "r-cpu.3-6.m54",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.3-6:m54"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
